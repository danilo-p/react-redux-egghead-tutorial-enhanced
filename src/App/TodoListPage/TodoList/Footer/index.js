import React from 'react';
import FilterLink from './FilterLink';
/**
 * Describes the footer of the app.
 *
 * @returns {jsx} The Footer template
 */
const Footer = () => (
  <p>
    Show:&nbsp;&nbsp;
    <FilterLink filter="SHOW_ALL">All</FilterLink>
    <FilterLink filter="SHOW_ACTIVE">Active</FilterLink>
    <FilterLink filter="SHOW_COMPLETED">Completed</FilterLink>
  </p>
);

export default Footer;
