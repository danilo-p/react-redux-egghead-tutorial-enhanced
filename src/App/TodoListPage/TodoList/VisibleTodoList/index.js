import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { toggleTodo } from 'common/ducks/todos';
import Todo from './Todo/index';

/**
 * Describes the todo list
 *
 * @param {any} props The component properties
 * @param {Array} props.todosList The array containing all the todos.
 * @param {Function} props.onTodoClick The todo click handler
 * @returns {jsx} The component template
 */
const VisibleTodoList = ({
  todosList,
  onTodoClick,
}) => (
  <ul>
    {todosList.map(t => (
      <Todo
        key={t.id}
        {...t}
        onClick={() => onTodoClick(t.id)}
      />
    ))}
  </ul>
);

/**
 * Utility for filtering an array of todos given a filter.
 *
 * @param {Array} todosList The todos list
 * @param {String} filter The filter to be applied on the todosList array.
 * @returns {jsx} The filtered array
 */
const getVisibleTodos = (
  todosList,
  filter,
) => {
  switch (filter) {
  case 'SHOW_COMPLETED':
    return todosList.filter(t => t.completed);
  case 'SHOW_ACTIVE':
    return todosList.filter(t => !t.completed);
  case 'SHOW_ALL':
  default:
    return todosList;
  }
};

VisibleTodoList.propTypes = {
  todosList: PropTypes.arrayOf(PropTypes.object).isRequired,
  onTodoClick: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  todosList: getVisibleTodos(
    state.todos,
    state.visibilityFilter,
  ),
});

const mapDispatchToProps = dispatch => ({
  onTodoClick: id => ({
    dispatch: dispatch(toggleTodo(id)),
  }),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(VisibleTodoList);
