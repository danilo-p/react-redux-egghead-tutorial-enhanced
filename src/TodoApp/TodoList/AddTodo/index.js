import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Input, Button } from 'semantic-ui-react';
import { addTodo } from 'common/ducks/todos';
import styles from './index.styl';

/**
 * Used for add a new todo with an input.
 *
 * @class AddTodo
 * @extends {React.Component}
 *
*/
class AddTodo extends React.Component {
  /**
   * Creates an instance of AddTodo.
   * @memberof AddTodo
   */
  constructor() {
    super();
    this.state = {
      inputText: '',
    };
    this.onInputChange = this.onInputChange.bind(this);
    this.onButtonClick = this.onButtonClick.bind(this);
  }

  /**
   * Handles the input change and updates the component state.
   *
   * @param {any} event The change event containing the current value of the
   * input.
   * @memberof AddTodo
   * @return {void}
   */
  onInputChange(event) {
    this.setState({
      inputText: event.target.value,
    });
  }
  /**
   * Handles the button click and dispatch an addTodo action if allowed.
   *
   * @memberof AddTodo
   * @return {void}
   */
  onButtonClick() {
    const { inputText } = this.state;
    if (inputText) {
      this.props.dispatch(addTodo(inputText));
      this.setState({
        inputText: '',
      });
    }
  }
  /**
   * Renders the component
   *
   * @memberof AddTodo
   * @return {jsx} The component template
   */
  render() {
    return (
      <div className="AddTodo">
        <Input
          action
          placeholder="Remind to feed the..."
          className={styles.AddTodo__Input}
          value={this.state.inputText}
          onChange={this.onInputChange}
          onKeyPress={(event) => {
            if (event.key === 'Enter') {
              this.onButtonClick();
            }
          }}
        >
          <input />
          <Button
            primary
            onClick={this.onButtonClick}
            className="AddTodo__Button"
          >
            Add Todo
          </Button>
        </Input>
      </div>
    );
  }
}

AddTodo.propTypes = {
  dispatch: PropTypes.func.isRequired,
};

export default connect()(AddTodo);
