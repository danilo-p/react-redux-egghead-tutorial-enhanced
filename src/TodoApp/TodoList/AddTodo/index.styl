.AddTodo__Input
    margin-right: 7.5px;
