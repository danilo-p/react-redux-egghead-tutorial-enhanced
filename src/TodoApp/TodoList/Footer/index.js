import React from 'react';
import FilterButton from './FilterButton';
/**
 * Describes the footer of the app.
 *
 * @returns {jsx} The Footer template
 */
const Footer = () => (
  <p>
    Show:&nbsp;&nbsp;
    <FilterButton filter="SHOW_ALL">All</FilterButton>
    <FilterButton filter="SHOW_ACTIVE">Active</FilterButton>
    <FilterButton filter="SHOW_COMPLETED">Completed</FilterButton>
  </p>
);

export default Footer;
