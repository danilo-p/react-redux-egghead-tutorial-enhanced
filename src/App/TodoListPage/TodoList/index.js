import React from 'react';
import AddTodo from './AddTodo';
import VisibleTodoList from './VisibleTodoList';
import Footer from './Footer';

/**
 * Describes the todo list.
 *
 * @return {jsx} The TodoList template
 */
export default function TodoList() {
  return (
    <div>
      <AddTodo />
      <VisibleTodoList />
      <Footer />
    </div>
  );
}
