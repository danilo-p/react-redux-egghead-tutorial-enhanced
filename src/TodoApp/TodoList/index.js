import React from 'react';
import AddTodo from './AddTodo';
import VisibleTodoList from './VisibleTodoList';
import Footer from './Footer';

/**
 * SFC that describes the todo list.
 * @see SFC https://reactjs.org/docs/components-and-props.html#functional-and-class-components
 *
 * @export
 * @return {jsx} The TodoList template
 */
export default function TodoList() {
  return (
    <div>
      <AddTodo />
      <VisibleTodoList />
      <Footer />
    </div>
  );
}
