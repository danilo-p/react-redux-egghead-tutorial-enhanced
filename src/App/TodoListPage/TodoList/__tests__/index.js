import React from 'react';
import basicComponentTests from 'tests/basicComponentTests';
import TodoList from '../';

describe('TodoList', () => {
  basicComponentTests(<TodoList />);
});
