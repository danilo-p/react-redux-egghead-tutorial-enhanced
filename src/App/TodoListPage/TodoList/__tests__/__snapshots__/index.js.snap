// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should render the component 1`] = `
<div>
  <Connect(AddTodo) />
  <Connect(VisibleTodoList) />
  <Footer />
</div>
`;
